stackrox/stackrox
# fix(ui): Fix accessibility issues with dl in Violations Details tab

## Description

### Problems

Almost 100 serious issues from axe DevTools

> dl elements must must only directly contain … elements

> dt and dd must be contained in a dl

### Analysis

1. Although guideline allows one `div` (that is, PatternFly `DescriptionGroup` element) between `dl` and descendant `dt` and `dd` elements, nested description lists have an additional `div` element.
2. Elsewhere PatternFly `Divider` element are another cause of addtional `div` element.

### Solution

Replace generic `ObjectDescriptionList` component with specific components that we know do not need extra `div` element:
1. `FlatObjectDescriptionList` for annotations and labels which have `Record<string, string>` type.
2. `PortDescriptionList` for `PortConfig` property of deployment.
3. Refactor plural comonents to remove need for `Divider` element:
    * Replace `ContainerSecrets` with `ContainerSecretDescriptionList` component.
    * Replace `ContainerVolumes` with `ContainerVolumeDescriptionList` component.
4. Import types from frontend source of truth. Rename components to prevent collision with response types.

### Residue

1. **Dave** can you take a look at last accessibility issue?

    > Certain ARIA roles must contain particular children

    ```html
    <table role="grid" class="pf-c-table pf-m-grid-md pf-m-compact" data-ouia-component-type="PF4/Table" data-ouia-safe="true" data-ouia-component-id="OUIA-Generated-Table-1">
    ```

    Related node

    ```html
    <h3 data-ouia-component-type="PF4/Title" data-ouia-safe="true" data-ouia-component-id="OUIA-Generated-Title-10" class="pf-c-title pf-m-lg">All network policies</h3>
    ```

2. Move `h3` headings into cards to solve serious accessibility problem, although not reported by axe DevTools: that screen reader navigation by heading is futile. Whenever elements have `aria-label` attribute that is redundant with visible text, it is hint of accessibility problems.

3. Thinking outside the box, investigate whether multiple containers present similar information about deployments.

## Checklist

## Testing Performed

Small net change because replaced generic `ObjectDescriptionList` with container-specific components.

1. `yarn lint` in ui
2. `yarn build` in ui and then compute branch - master for the following
    * `wc build/static/js/*.js`
        main.css: -4 = 3758679 - 3758683
        total: 108 = 9930671 - 9930563
    * `ls -al build/static/js/*.js | wc -l`
        0 = 81 - 81 files

### Manual testing

1. Visit /main/violations, click a link, and then click **Deployment**

    * See at right explicit **containers[0]** heading and also **volumes[0]** and **volumes[1]** terms
        ![violations_deployment_1](https://github.com/stackrox/stackrox/assets/11862657/ebe3a887-3e51-411f-b2f5-46d12c7241df)

    * See at left explicit **ports[0]** and **ports[1]** and **ports[2]** headings
        ![violations_deployment_2](https://github.com/stackrox/stackrox/assets/11862657/32897c81-b210-4738-b91e-06ff7b6e0de6)

    * See at right **secrets** and **None** in contrast to **secrets[0]** with **Name** and **Container path** above
        ![violations_deployment_3](https://github.com/stackrox/stackrox/assets/11862657/37047a75-b363-46bd-a4e7-31522206a627)

### Integration testing

* violations/violations.test.js
